source/_posts/2016-01-05-improvements-to-error-responses-in-rails-5-api-mode.markdown

```diff
@@ -0,0 +1,160 @@
+{"status":"400","error":"Bad Request"}
+```
+
+The difference could be very subtle, but if our client component makes any further processing based on the response content in case of errors, we may need to use the `Accept` header to be safe, if we use Rails 4.2 or previous versions.
+
+As part of our changes related with error responses on Rails, we added some code to respond with the format indicated with the request path when an exception was raised before executing the router code.
+The approach is very simple: we retrieve the format from the extension of the URL _guessing_ that the format is the extension part of the request path.
+
+This way to determine the format is not always equivalent to what is being parsed by the router, because the `:format` fragment could theorically be in any part of the path. However, we think that will work for all but a very few cases. In fact, I've never seen a Rails application with custom routes where the `:format` fragment is not the extension part of the URL.
+
+You can see what was finally changed for Rails 5 by looking at the `MimeNegotiation#format_from_path_extension` method [here](https://github.com/rails/rails/blob/fc027e1faf39f0deacd734fe0a46eb5ee6d70851/actionpack/lib/action_dispatch/http/mime_negotiation.rb#L166-L171). Also, you can know more about the limitations of this approach when you have routes with constraints in this [discussion](https://github.com/rails/rails/issues/22747#issuecomment-166503385).
+
+
+## Error responses on development mode
+
+Up to now, we have discussed how error responses are formatted with the default production settings. In fact, there is a specific Rails configuration setting called `action_controller.consider_all_requests_local` that controls how errors are rendered. It's set to `false` by default on `config/environments/production.rb` causing that error responses present the problem in a user-friendly way (generally, an user-oriented error message without exposing internal information about the exception).
+
+In contrast, `consider_all_requests_local` is set to `true` by default on development mode, and it generates the standard Rails HTML error page for debugging that shows the exception message, the stack trace and also the [web console](https://github.com/rails/web-console) where we can execute code on the server. However, you can not get an error response with JSON format on Rails 4.2, because it responds with this HTML error page, except for XHR requests where error responses are rendered in plain text.
+
+However, with the arrival of the Rails API mode, we found that error responses behaving in this way, on development mode, could not be so convenient when you're working on a Rails API project. In many cases, you may want to receive error responses in the format expected by your client component. In this case, the error responses should be formatted in JSON, as it's on production mode, but also including information about the error for debugging purposes.
```

Comment:
This paragraph is a bit confusing. First talks about about development mode, and finally switches to production mode saying that it should include information for debugging purposes. I guess you meant development in the ending part?.